### DISCUSSION TITLE
JSLint Node issues when building 3.0.1 in docker

### DISCUSSION THREAD

**Discussion by FatConan** [playframework/playframework#12291] [Q&A]
Hi,

Bit of a niche question this, but has anyone encountered any node issues around jshint since upgrading to 3.0.1 and running the build inside docker? 

For context: at the end of last year I wrote a gitlab CI script that built a fairly big Java play project (at that time version 2.8.21) inside a docker container using openJDK11 and ubuntu 22.04, which all worked well. This year I decided to upgrade to play 3.0.1 and openJDK21, it's all gone well and builds/runs perfectly on my laptop, desktop (both ubuntu 22.04, both openJDK21), but it now always fails in docker (set up with the same OS, java version, nodejs version etc.):

```
[info]   __              __
[info]   \ \     ____   / /____ _ __  __
[info]    \ \   / __ \ / // __ `// / / /
[info]    / /  / /_/ // // /_/ // /_/ /
[info]   /_/  / .___//_/ \__,_/ \__, /
[info]       /_/               /____/
[info] 
[info] Version 3.0.1 running Java 21.0.1
[info] 
[info] Play is run entirely by the community. Please consider contributing and/or donating:
[info] https://www.playframework.com/sponsors
[info] 
[success] Total time: 0 s, completed Jan 10, 2024, 10:07:08?AM
[success] Total time: 19 s, completed Jan 10, 2024, 10:07:27?AM
[info] compiling 515 Scala sources and 1184 Java sources to /builds/xxx/yyy/target/scala-2.13/classes ...
[info] LESS compiling on 27 source(s)
[info] JavaScript linting on 392 source(s)
[error] internal/modules/cjs/loader.js:818
[error]   throw err;
[error]   ^
[error] Error: Cannot find module 'strip-json-comments'
[error] Require stack:
[error] - /builds/xxx/yyy/project/target/jshint/jshint-shell.js
[error]     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)
[error]     at Function.Module._load (internal/modules/cjs/loader.js:667:27)
[error]     at Module.require (internal/modules/cjs/loader.js:887:19)
[error]     at require (internal/modules/cjs/helpers.js:74:18)
[error]     at /builds/xxx/yyy/project/target/jshint/jshint-shell.js:25:29
[error]     at Object.<anonymous> (/builds/xxx/yyy/project/target/jshint/jshint-shell.js:74:2)
[error]     at Module._compile (internal/modules/cjs/loader.js:999:30)
[error]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
[error]     at Module.load (internal/modules/cjs/loader.js:863:32)
[error]     at Function.Module._load (internal/modules/cjs/loader.js:708:14) {
[error]   code: 'MODULE_NOT_FOUND',
[error]   requireStack: [
[error]     '/builds/xxx/yyy/project/target/jshint/jshint-shell.js'
[error]   ]
[error] }
```

I've tried installing different versions of node, I've tried explicitly installing the missing module using npm, and though I can generate a different node error (like complaining that `[error] Error [ERR_REQUIRE_ESM]: require() of ES Module /root/yyy/node_modules/strip-json-comments/index.js from /root/yyy/project/target/jshint/jshint-shell.js not supported.`) I haven't been able to get past it. 

I can't work out what the problematic difference between my laptop, desktop, the previous docker version (all of which were running the bog standard nodejs and npm installations from apt) and this docker container environment could be. If anyone has any suggestions, or can see something I've stupidly overlooked, that would be greatly appreciated.

**Comment by mkurz** [comment]
Are you 100% sure the node and npm version are exactly the same on your desktop/laptop and the docker container?
Please check in your container and on your machine(s):
```
node --version
npm --version
```

**Reply by FatConan** [reply]
It turns out that neither my laptop nor desktop has npm installed (I must have failed to reinstall it when I upgraded to 22.04) so I removed that apt call from the docker definition and retested: still suffered the same problem.  All three are using the same version of node, and none now have npm installed.

*Laptop*
```
$ node --version
v12.22.9
```
*Desktop*
```
$ node --version
v12.22.9
```
*Docker*
```
# node --version
v12.22.9
```

**Reply by mkurz** [reply]
Not having `npm` installed might be a problem...
Can you run `whereis node` everywhere? This will give you the location where node is installed. Now, can you check if in the same folder there is also a `npm` command? 
Something like:
```sh
$ whereis node
node: /home/mkurz/.nvm/versions/node/v20.5.1/bin/node

# Then check if there is a npm command inside the same folder:
$ ls -l /home/mkurz/.nvm/versions/node/v20.5.1/bin/npm

**Reply by mkurz** [reply]
...and if there is a command inside then please run it with the `--version` flag:
 ```sh
/home/mkurz/.nvm/versions/node/v20.5.1/bin/npm --version
```

**Reply by mkurz** [reply]
I guess you will find there is no `npm` command, so very likely sbt-js-engine, which Play is using is falling back to very old npm version provided by webjars.

Also I see that the node version provided by Ubuntu 22.04 LTS is very outdated (version 12): https://packages.ubuntu.com/jammy/nodejs

I am pretty sure you will get problems using such an old node version and not having npm installed.

What I would do:
Remove that old node version from all your systems and switch to [nvm, the node version manager](https://github.com/nvm-sh/nvm) to install and manage node and npm version installed on your machines:
```
sudo apt purge nodejs
```
You find guides on how to install and use nvm here:
* https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-22-04#option-3-installing-node-using-the-node-version-manager
* https://tecadmin.net/how-to-install-nvm-on-ubuntu-22-04/

I highly recommend to use one of the latest node LTS versions, which is either 20 or 18, but nothing below.
Also, good thing about nvm is, that when you switch the node version, it automaticially installs and switched npm as well, using a version that is compatible with the currently in use node version.

> The correct version of Node is installed on our machine as we expected. A compatible version of npm is also available.

**Reply by mkurz** [reply]
...just make sure then that the correct/desired `node` and `npm` versions are used, by checking with the `--version` flag, just to be save.

**Comment by FatConan** [comment]
Hi, thanks for the help. We think we've managed to get it to work, but the configuration for docker still seems a little different to that required outside of it, and as yet we haven't been able to work out why. Hopefully we can nail down what's causing the issue for us (if it isn't just somehow illusory) and I can write it up here in case anyone else encounters it.
